bip-0324.mediawiki

```diff
@@ -0,0 +1,467 @@
+</pre>
+
+Once a peer has received the public key from its counterparty, the shared secret MUST be calculated by using secp256k1 ECDH.
+
+Private keys will never be transmitted. The shared secret can only be calculated if an attacker knows at least one private key and the counterparty's public key. This key-exchange is based on the discrete log problem and thus not sufficiently strong against known forms of possible quantum computer algorithms. Adding an additional quantum resistant key exchange like NewHope is possible but out of scope for this proposal.
+
+After a successful handshake, messages MUST use the "v2 messages structure". Non-encrypted v1 messages from the initiating peer MUST lead to an immediate connection termination.
+
+After a successful handshake, both peers MUST wipe the ephemeral-session-key from memory and/or persistence storage.
+
+A peer not supporting this proposal will not perform the described handshake and thus send a v1 version message.
+Peers supporting this BIP MAY optionally allow unencrypted v1 communication by detecting a v1 version message by the initial 11-byte sequence of <code>4byte net magic || "version"</code>.
+
+=== Symmetric Encryption Cipher Keys ===
+
+Once the ECDH secret (<code>ECDH_KEY</code>) is calculated on each side, the symmetric encryption cipher keys MUST be derived with HKDF <ref>[https://tools.ietf.org/html/rfc5869 HKDF (RFC 5869)]</ref> after the following specification:
+
+1. HKDF extraction
+<code>PRK = HKDF_EXTRACT(hash=SHA256, salt="BitcoinSharedSecret||INITIATOR_32BYTES_PUBKEY||RESPONDER_32BYTES_PUBKEY||NETWORK_MAGIC", ikm=ECDH_KEY)</code>.
+
+2. Derive Key_1_A (K_1 communication direction A)
+<code>K1A = HKDF_EXPAND(prk=PRK, hash=SHA256, info="BitcoinK_1_A", L=32)</code>
+
+2. Derive Key_2_A (K_2 communication direction A)
+<code>K1B = HKDF_EXPAND(prk=PRK, hash=SHA256, info="BitcoinK_2_A", L=32)</code>
+
+3. Derive Key_1_B (K_1 communication direction B)
+<code>K2 = HKDF_EXPAND(prk=PRK, hash=SHA256, info="BitcoinK_1_B", L=32)</code>
+
+3. Derive Key_2_B (K_2 communication direction B)
+<code>K2 = HKDF_EXPAND(prk=PRK, hash=SHA256, info="BitcoinK_2_B", L=32)</code>
+
+=== Session ID ===
+
+Both parties MUST also calculate the 256bit session-id using <code>SID = HKDF_EXPAND(prk=PRK, hash=SHA256, info="BitcoinSessionID", L=32)</code>. The session-id can be used for authenticating the encryption-session (identity check).
+
+The session-id MUST be presented to the user on request.
+
+=== ChaCha20-Poly1305@Bitcoin Cipher Suite ===
+
+==== Background ====
+
+ChaCha20-Poly1305@Bitcoin AEAD is almost identical to the openSSH version <ref>[https://github.com/jhcloos/openssh-chacha-poly1305/blob/master/PROTOCOL.chacha20poly1305]</ref>. The only difference to the @openSSH version is that the @Bitcoin version has a 3 byte package length (instead 4) and reuses the remaining bytes of the ChaCha20 round (21 length field encryption per ChaCha20 round; 21 x 3 < 64).
```

Comment:
openSSH -> OpenSSH.

What does the ampersand mean?

I think the x is multiplication, but I'd suggest just spelling it out instead: "21 times 3 is less than 64".